paasta_tools/cli/cmds/status.py

```diff
@@ -1325,35 +1388,35 @@ def get_replica_state(pod: KubernetesPodV2) -> ReplicaState:
         if not pod.containers or all([c.state == "waiting" for c in pod.containers]):
             state = ReplicaState.STARTING
         else:
-            state = ReplicaState.UNHEALTHY
+            state = ReplicaState.NOT_RUNNING
     elif phase == "Running":
-        if not pod.ready:
-            # TODO: Take sidecar containers into account
-            #   This logic likely needs refining
-            state = ReplicaState.UNREACHABLE
-            main_container = [
-                c
-                for c in pod.containers
-                if c.name not in kubernetes_tools.SIDECAR_CONTAINER_NAMES
-            ]
-            if main_container:
-                if main_container[0].restart_count > 0:
-                    if (
-                        pod.create_timestamp
-                        + main_container[0].healthcheck_grace_period
-                        > datetime.utcnow().timestamp()
-                    ):
-                        state = ReplicaState.WARMING_UP
-                    else:
-                        state = ReplicaState.HEALTHCHECK_FAILING
-                else:
-                    # No restarts yet, still warming up?
-                    state = ReplicaState.WARMING_UP
+        ####
+        #
+        # TODO: Take sidecar containers into account
+        #   This logic likely needs refining
+        main_container = get_main_container(pod)
+        if main_container:
+            warming_up = (
+                pod.create_timestamp + main_container.healthcheck_grace_period
+                > datetime.now().timestamp()
+            )
+            if main_container.state != "running" or not pod.ready:
+                state = ReplicaState.NOT_RUNNING
+            elif pod.mesh_ready is False and not warming_up:
+                state = ReplicaState.UNREACHABLE
+            elif recent_liveness_failure(pod) or recent_container_restart(
+                main_container
+            ):
+                state = ReplicaState.WARNING
+            elif warming_up:
```

Comment:
I think this means that any pod that would otherwise be RUNNING (is ready, in the running state, no recent warnings) would always be in the WARMING_STATE until the grace period is over.

I would suggest checking the warming_up state underneath each condition where the pod isn't ready instead (e.g. under `not pod.ready` and `pod.mesh_ready is False`). I think that will make the logic for the warming up state more explicit.

Reply:
To sum up the 2 comments on this function, I think it might be easier to reason about if the code were structured like the following, to make the branches for each possible state more obvious. 

```
if not_running_state_a:
  if state_a_reason_one:
    return X
  ...
  else:
    return Y
elif not_running_state_b:
  ...
else:
  return RUNNING
```